pkg/proxy/logger/logger.go

```diff
@@ -232,18 +232,19 @@ func (lr *LogRecord) Log() {
 	flowdebug.Log(lr.getLogFields(), "Logging flow record")
 
 	logMutex.Lock()
-	defer logMutex.Unlock()
-
 	lr.Metadata = metadata
+	n := notifier
+	logMutex.Unlock()
 
-	if notifier != nil {
-		notifier.NewProxyLogRecord(lr)
+	if n != nil {
+		n.NewProxyLogRecord(lr)
 	}
 }
 
 // LogRecordNotifier is the interface to implement LogRecord notifications
 type LogRecordNotifier interface {
-	// NewProxyLogRecord is called for each new log record
+	// NewProxyLogRecord is called for each new log record.
+	// It may be called concurrently by multiple logger goroutines.
 	NewProxyLogRecord(l *LogRecord) error
```

Comment:
I think you can go a bit further here and say that any object which wishes to implement the `LogRecordNotifier` interface, must support concurrent calls to the interface methods, and not rely on locks.